homeassistant/components/sensor/elkm1.py

```diff
@@ -0,0 +1,315 @@
+    return temperature if temperature > undefined_temperature else None
+
+
+class ElkSensor(ElkEntity):
+    """Base representation of Elk-M1 sensor."""
+
+    def __init__(self, element, elk, elk_data):
+        """Initialize the base of all Elk sensors."""
+        super().__init__(element, elk, elk_data)
+        self._state = None
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+
+class ElkCounter(ElkSensor):
+    """Representation of an Elk-M1 Counter."""
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return 'mdi:numeric'
+
+    def _element_changed(self, element, changeset):
+        self._state = self._element.value
+
+
+class ElkKeypad(ElkSensor):
+    """Representation of an Elk-M1 Keypad."""
+
+    @property
+    def temperature_unit(self):
+        """Return the temperature unit."""
+        return self._temperature_unit
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return self._temperature_unit
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return 'mdi:thermometer-lines'
+
+    @property
+    def device_state_attributes(self):
+        """Attributes of the sensor."""
+        from elkm1_lib.util import username
+
+        attrs = self.initial_attrs()
+        attrs['area'] = self._element.area + 1
+        attrs['temperature'] = self._element.temperature
+        attrs['last_user_time'] = self._element.last_user_time.isoformat()
+        attrs['last_user'] = self._element.last_user + 1
+        attrs['code'] = self._element.code
+        attrs['last_user_name'] = username(self._elk, self._element.last_user)
+        return attrs
+
+    def _element_changed(self, element, changeset):
+        self._state = temperature_to_state(self._element.temperature, -40)
+
+    async def async_added_to_hass(self):
+        """Register callback for ElkM1 changes and update entity state."""
+        await super().async_added_to_hass()
+        self.hass.data[ELK_DOMAIN]['keypads'][
+            self._element.index] = self.entity_id
+
+
+class ElkPanel(ElkSensor):
+    """Representation of an Elk-M1 Panel."""
+
+    async def async_added_to_hass(self):
+        """Register callback for ElkM1 changes."""
+        await super().async_added_to_hass()
+        async_dispatcher_connect(
+            self.hass, '{}_{}'.format(SIGNAL_SPEAK_WORD, self.entity_id),
+            self._speak_word)
+        async_dispatcher_connect(
+            self.hass, '{}_{}'.format(SIGNAL_SPEAK_PHRASE, self.entity_id),
+            self._speak_phrase)
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return "mdi:home"
+
+    @property
+    def device_state_attributes(self):
+        """Attributes of the sensor."""
+        attrs = self.initial_attrs()
+        attrs['elkm1_version'] = self._element.elkm1_version
```

Comment:
Firmware version and similar info should go in device info, not state attributes. State attributes should hold data connected to the state of the device. `device_info` property requires config entries though so it's not super easy to implement. I suggest removing the version attributes for now.

https://developers.home-assistant.io/docs/en/device_registry_index.html

Reply:
I've read those docs and was trying to figure out how it could be used with ElkM1. Not very clear at all how to do that integration. Some examples would be helpful in the docs. `config_entries` not very clear at all either. Does that mean everything that's in the `elk` section of `configuration.yaml` goes away?

So what to do? I'd like to keep the versions, implementing something else way over the top (to use your words "not super easy to implement").

What about the state associated with this entity? It tracks the state of the connection from hass to the actual platform. Is that in `device_info`?

Reply:
We make a distinction between dynamic info and static info. Dynamic info belongs in state and state attributes as long as it pertains to the state of a device. Static info is... static so doesn't have to do with state of the device.

Static info like version has to be removed from state attributes.

If the connection has multiple states it can be in, then it fits as sensor state.